internal/util/cmdutil/cmdutil.go

```diff
@@ -194,67 +197,87 @@ func GetKeywordsFromFlag(cmd *cobra.Command) []string {
 // SuggestFunctions looks for functions from kpt curated catalog list as well as the Porch
 // orchestrator to suggest functions.
 func SuggestFunctions(cmd *cobra.Command) []string {
-	var functions []string
-	matchers := []porch.Matcher{
-		porch.TypeMatcher{FnType: cmd.Flag("type").Value.String()},
-		porch.KeywordsMatcher{Keywords: GetKeywordsFromFlag(cmd)},
-	}
-	porchFunctions := porch.MatchFunctions(cmd.Context(), matchers...)
-	functions = append(functions, porch.ToShortNames(porchFunctions)...)
-	// TODO(yuwenma): We should filter these catalog.kpt.fn functions by its annotation as well! This requires changing
-	// the kpt-functions-catalog release workflow to include new annotations.
-	functions = append(functions, FetchFunctionImages()...)
-	var dedup []string
-	fnMap := map[string]bool{}
-	for _, function := range functions {
-		if _, ok := fnMap[function]; !ok {
-			dedup = append(dedup, function)
-			fnMap[function] = true
-		}
+	matchers := []function.Matcher{
+		function.TypeMatcher{FnType: cmd.Flag("type").Value.String()},
```

Comment:
I wonder do you validate the --type flag value somewhere?

Reply:
Ah, you catch it. Short answer: no.  
The behavior of default value is something that I haven't come up with a good answer yet (well, and maybe not have put enough thoughts on it):

Problem 1. 
`kpt fn eval` already has too many flags that sort of correlate to each other. `--keywords`, `--type`, `--exec`, `--image`. The usage of `--type` is determined and determines other behavior. I don't want to give users too much burden to understand all of these. So it may not be a good idea to use --type flag.

Problem 2. 
`--type` only really matters when used with "--save". At which point "mutator" is a valid default. But when used as a filter to functions, I'm not quite confident if we shall use it as a filter. 

Maybe the right approach should be:
-  `--type` is default to "" (rather than "mutator") with autocompletion enabled to suggest "validator" and "mutator".
- Matching functions by type accepts the "--type" to be empty string "", in such case suggest all functions. 
- Only when "--save" is true, will kpt validate "--type". It accepts value `["", "mutator", "validator"]`, and reassign `""` to `"mutator"`.

The advantage is that users are guided to only care the `--type` when they want to write function to Kptfile `pipeline.validators`. All other cases the `--type` is a nice to have thing.  

My main concern here is that these assumptions could further confuse users. WDYT? @mengqiy

Reply:
I filed https://github.com/GoogleContainerTools/kpt/issues/3014 for further discussion about this flag. I'm going to merge this PR.